File: java/src/kg/KnapsackSolver.java
package kg;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class KnapsackSolver {
    public KnapsackResult solve(List<KnapsackItem> items, int capacity) {
        int itemCount = items.size();
        int[][] dp = new int[itemCount + 1][capacity + 1];

        for (int i = 1; i <= itemCount; i++) {
            KnapsackItem item = items.get(i - 1);
            for (int w = 0; w <= capacity; w++) {
                dp[i][w] = dp[i - 1][w];
                if (item.getWeight() <= w) {
                    dp[i][w] = Math.max(dp[i][w], dp[i - 1][w - item.getWeight()] + item.getValue());
                }
            }
        }

        List<KnapsackItem> selected = new ArrayList<>();
        int remainingCapacity = capacity;
        int totalWeight = 0;
        for (int i = itemCount; i >= 1; i--) {
            if (dp[i][remainingCapacity] != dp[i - 1][remainingCapacity]) {
                KnapsackItem item = items.get(i - 1);
                selected.add(item);
                remainingCapacity -= item.getWeight();
                totalWeight += item.getWeight();
            }
        }

        Collections.reverse(selected);
        return new KnapsackResult(selected, dp[itemCount][capacity], totalWeight);
    }
}

File: python/solver/knapsack_solver.py
from .models import KnapsackItem


def solve_knapsack(items: list[KnapsackItem], capacity: int) -> dict:
    dp = [[0] * (capacity + 1) for _ in range(len(items) + 1)]
    for i, item in enumerate(items, start=1):
        for weight in range(capacity + 1):
            dp[i][weight] = dp[i - 1][weight]
            if item.weight <= weight:
                dp[i][weight] = max(dp[i][weight], dp[i - 1][weight - item.weight] + item.value)

    selected: list[KnapsackItem] = []
    remaining = capacity
    total_weight = 0
    for i in range(len(items), 0, -1):
        if dp[i][remaining] != dp[i - 1][remaining]:
            item = items[i - 1]
            selected.append(item)
            remaining -= item.weight
            total_weight += item.weight
    selected.reverse()
    return {"selected_items": selected, "total_value": dp[len(items)][capacity], "total_weight": total_weight}
